bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/cache/TTSLRUCacheImpl.java

```diff
@@ -0,0 +1,371 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.voice.internal.cache;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileStore;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.audio.AudioFormat;
+import org.openhab.core.audio.AudioStream;
+import org.openhab.core.voice.TTSException;
+import org.openhab.core.voice.TTSService;
+import org.openhab.core.voice.Voice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Cache system to avoid requesting {@link TTSService} for the same utterances.
+ * This is a LRU cache (least recently used entry is evicted if the size
+ * is exceeded)
+ * Size is based on the size on disk (in kilobytes)
+ *
+ * @author Gwendal Roulleau - Initial contribution
+ */
+@NonNullByDefault
+public class TTSLRUCacheImpl implements TTSCache {
+
+    private final Logger logger = LoggerFactory.getLogger(TTSLRUCacheImpl.class);
+
+    public static final String INFO_EXT = ".info";
+    public static final String SOUND_EXT = ".snd";
+
+    private static final String VOICE_TTS_CACHE_PID = "org.openhab.voice.tts";
+    private static final String CACHE_FOLDER_NAME = "cache";
+
+    /**
+     * A map to store all cached entries
+     */
+    private Map<String, TTSResult> ttsResultMap;
+
+    /**
+     * The most recently used {@link TTSResult}
+     */
+    @Nullable
+    private TTSResult head;
+
+    /**
+     * The least recently used {@link TTSResult}. Could be evicted soon.
+     */
+    @Nullable
+    private TTSResult tail;
+
+    /**
+     * Lock the cache to handle concurrency
+     */
+    private ReentrantLock lock = new ReentrantLock();
+
+    /**
+     * The size limit, in bytes. The size is not a hard one, because the final size of the
+     * current request is not known and may or may not exceed the limit.
+     */
+    private long limitSize;
+
+    private File cacheFolder;
+
+    /**
+     * Check free disk space and construct a cache system for TTS result.
+     *
+     * @param Target size of the cache in kB. The size is not a hard one, because the final size of the
+     *            current request is not known and may exceed the limit.
+     * @throws IOException when we cannot create the cache directory or if we have not enough space (*2 security margin)
+     */
+    public TTSLRUCacheImpl(long size) throws IOException {
+        this.ttsResultMap = new ConcurrentHashMap<>();
```

Comment:
Since you are implementing an LRU cache, wouldn't a `LinkedHashMap` not simplify your implementation a lot? You can configure it to order on last access and you can override `removeEldestEntry` with your own implementation that checks if the least recently used entry needs to be deleted, delete the file and then remove it from the collection.

Reply:
I was not aware of this functionnality of the LinkedHashMap.
It could have saved me some time !

I will see if I can use it, thanks !

Reply:
Done.
But with a catch : I couldn't use the automatic delete functionnality of the LinkedHashMap (usually done by overriding the `removeEldestEntry` method).
Indeed, the automatic delete function prune the oldest entry in the map if the `removeEldestEntry` returns true, but it does this only once by `put` action.
Or, in a limited size cache, it could be necessary to prune several small entries if the new entry is a big one.

So, I had to keep a custom `put` function that call `makeSpace`.